ui/raidboss/emulator/data/Encounter.ts

```diff
@@ -1,56 +1,78 @@
 import CombatantTracker from './CombatantTracker';
-import LogEventHandler from './LogEventHandler';
 import PetNamesByLang from '../../../../resources/pet_names';
 import EmulatorCommon from '../EmulatorCommon';
 import LogRepository from './network_log_converter/LogRepository';
 import NetworkLogConverter from './NetworkLogConverter';
+import { Lang } from 'types/global';
+import LineEvent, { isLineEventSource, isLineEventTarget } from './network_log_converter/LineEvent';
+import { UnreachableCode } from '../../../../resources/not_reached';
 
-const isPetName = (name, language = undefined) => {
+const isPetName = (name: string, language: Lang | undefined = undefined) => {
   if (language)
     return PetNamesByLang[language].includes(name);
 
   for (const lang in PetNamesByLang) {
-    if (PetNamesByLang[lang].includes(name))
+    if (PetNamesByLang[lang as Lang].includes(name))
       return true;
   }
 
   return false;
 };
 
+const validTimestamp = (timestamp: number) => {
+  return timestamp > 0 && timestamp < Number.MAX_SAFE_INTEGER;
+};
+
 export default class Encounter {
-  constructor(encounterDay, encounterZoneId, encounterZoneName, logLines) {
+  private static readonly encounterVersion = 1;
+  public id?: number;
+  version: number;
+  initialOffset = Number.MAX_SAFE_INTEGER;
+  endStatus = 'Unknown';
+  startStatus = 'Unknown';
+  private engageAt = Number.MAX_SAFE_INTEGER;
+  private firstPlayerAbility = Number.MAX_SAFE_INTEGER;
+  private firstEnemyAbility = Number.MAX_SAFE_INTEGER;
+  firstLineIndex = 0;
+  combatantTracker?: CombatantTracker;
+  startTimestamp = 0;
+  endTimestamp = 0;
+  duration = 0;
+  playbackOffset = 0;
+  language: Lang = 'en';
+
+  constructor(
+    public encounterDay: string,
+    public encounterZoneId: string,
+    public encounterZoneName: string,
+    public logLines: LineEvent[]) {
     this.version = Encounter.encounterVersion;
-    this.id = null;
-    this.encounterZoneId = encounterZoneId;
-    this.encounterZoneName = encounterZoneName;
-    this.encounterDay = encounterDay;
-    this.logLines = logLines;
   }
 
-  initialize() {
-    this.initialOffset = Number.MAX_SAFE_INTEGER;
-    this.endStatus = 'Unknown';
-    this.startStatus = new Set();
-    this.engageAt = Number.MAX_SAFE_INTEGER;
-    this.firstPlayerAbility = Number.MAX_SAFE_INTEGER;
-    this.firstEnemyAbility = Number.MAX_SAFE_INTEGER;
-
-    this.firstLineIndex = 0;
+  initialize(): void {
+    const startStatuses = new Set<string>();
 
     for (let i = 0; i < this.logLines.length; ++i) {
       const line = this.logLines[i];
+      if (!line)
+        throw new UnreachableCode();
```

Comment:
For this style of for loop where you also need the index, you could also consider something like this to appease TypeScript:

```typescript
this.logLines.forEach((line, i) => {
```